### PR TITLE
Drop python 2 support

### PR DESCRIPTION
**Pull request by Sinclert**
This PR is intended to drop support for Python 2 starting at **version 0.8.0**, requiring Python 3.6+.

It was decided in issue https://github.com/diana-hep/madminer/issues/437 that the _Python-3-only_ codebase will coexist with the _master_ version of the project for now, in a new branch called `python3-only`.

## Context

As a reminder, Python 2 reached its _End Of Life_ back in January 1st 2020, and the latest version of PIP will drop its support at the beginning of 2021 ([reference](https://pip.pypa.io/en/latest/development/release-process/#python-2-support)). Therefore, one of this refactor goals is to avoid obsolescence of Python2 features due to lack of maintenance.

In addition, this project builds a Docker image which is used as based image for _down-stream_ repositories such as:

- [madminer-jupyter-env](https://github.com/scailfin/madminer-jupyter-env/blob/master/Dockerfile#L19).
- [madminer-workflow-ph](https://github.com/scailfin/madminer-workflow-ph)
- [madminer-workflow-ml](https://github.com/scailfin/madminer-workflow-ml).

By enforcing users to drop Python2 and only use Python3, we skip the installation of `madminer` on a Python2 environment, saving **up to 800 MBs per Docker image** (mostly due to [PyTorch installation](https://pytorch.org/get-started/locally/) on Linux).

## Changes

Codebase changes can mostly be grouped into 3 categories:

#### 1. Introduction of f-strings
Python 3.6 introduced [f-strings](https://www.python.org/dev/peps/pep-0498/), a shorter and [faster way](https://realpython.com/python-f-strings/#speed) to format strings using local variables. In this refactor f-strings have been introduced in two different scenarios:

A) When the `"...".format()` syntax was used to interpolate strings (exceptions found in the [mg.py](https://github.com/Sinclert/madminer/blob/drop-python-2/madminer/utils/interfaces/mg.py) module).
B) When the `<var> + "..."` syntax was used to concatenate strings.

#### 2. Removal of Python2 dependencies
The codebase had two dependencies targeting _Python2-Python3_ coexistence: `six` and `future`. Both of them have been removed from the package requirements, and all their references erased from the package codebase.

#### 3. Improve formatting
Finally, given that a thoughtful scan of the codebase was given, several typos were corrected, and several coding style guides were informally applied (derived from the [Python black](https://black.readthedocs.io/en/stable/the_black_code_style.html) code style).

## Review strategy
Commit by commit.

Each commit addresses a particular file at the root of the project (`Dockerfile`, `setup.py`...), or a particular folder within `madminer`, where all the source files are located. Moreover, each commit has a prefix showing what it addreses:

- `package`.
- `conda`.
- `docker`.
- `travis`.
- `tests`.
- `docs`.
- `src`.

## Future work
This PR leaves 2 **unfinished** tasks that may better be addressed in a separate PR, given the size of this one:

- [ ] The update of the [Jupyter notebook examples](https://github.com/Sinclert/madminer/tree/drop-python-2/examples) to Python 3.6+ syntax.
- [ ] The enforcement of [Black style guide](https://black.readthedocs.io/en/stable/the_black_code_style.html) by a [pre-commit](https://pre-commit.com/) hook / CI step.

### CODE REVIEW

File: madminer/delphes/delphes_reader.py [hunk 1]
```diff
@@ -814,78 +859,6 @@ def _analyse_delphes_sample(
...
-                logger.warning(
-                    "Observations for observable %s have non-numeric dtype %s. This usually means something "
-                    "is wrong in the definition of the observable. Data: %s",
-                    key,
-                    obs.dtype,
-                    obs,
-                )
-        return n_events
-
-    def _check_sample_weights(self, n_events, this_weights):
```
Comment (johannbrehmer) [inline madminer/delphes/delphes_reader.py:865]: What happened to this function, was it superfluous, or did you move the functionality somewhere else? Or am I just blind?

File: madminer/limits/asymptotic_limits.py [hunk 2]
```diff
@@ -84,7 +82,7 @@ def observed_limits(
         `x_observed` specifies the observed data as an array of observables, using the same observables and their order
         as used throughout the MadMiner workflow.
 
-        The p-values with frequentist hypothesis tests using the likelihood ratio as test statistic. The asymptotic
```
Comment (johannbrehmer) [inline madminer/limits/asymptotic_limits.py:87]: ["Frequentist"](https://en.wikipedia.org/wiki/Frequentist_inference) wasn't just a typo for "frequent" :) Could you revert this change, please?

File: madminer/delphes/delphes_reader.py (hunk 1 above)
Comment (Sinclert) [reply madminer/delphes/delphes_reader.py:865]: Both `_check_sample_observations` and `_check_sample_weights` functions are identical but for what they check (_observation_ vs _weight_).

I created a generic function at the beginning of the class called `_check_sample_elements`, that replaces these two.

Although this is not a pure _"Drop Python2"_ change, I thought it was an _easy win_. Please, let me know if you think this generic function should have any other name.

File: madminer/limits/asymptotic_limits.py (hunk 2 above)
Comment (Sinclert) [reply madminer/limits/asymptotic_limits.py:87]: Yes. My bad

File: madminer/utils/interfaces/mg.py [hunk 3]
```diff
@@ -507,12 +491,9 @@ def setup_mg_reweighting_with_scripts(
 
     #  Card copying commands
     if reweight_card_file_from_mgprocdir is not None:
-        copy_commands = "cp {}/{} {}{}\n".format(
-            mg_process_directory_placeholder,
-            reweight_card_file_from_mgprocdir,
-            mg_process_directory_placeholder,
-            "/Cards/reweight_card.dat",
-        )
+        copy_commands = f"cp " \
```
Comment (johannbrehmer) [inline madminer/utils/interfaces/mg.py:494]: Definitely better than before, but maybe this is even better:

```
copy_commands = (
    f"cp {mg_process_directory_placeholder}/{reweight_card_file_from_mgprocdir}"
    f" {mg_process_directory_placeholder}/Cards/reweight_card.dat\n"
)
```

[PEP8 says: "Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation."](https://www.python.org/dev/peps/pep-0008/)

File: madminer/utils/interfaces/mg.py [hunk 4]
```diff
@@ -536,9 +517,9 @@ def setup_mg_reweighting_with_scripts(
     make_file_executable(script_file)
 
     # How to call it from master script
-    call_instruction = "{}/{} [MG_process_directory] [log_directory]".format(
-        mg_process_directory, script_file_from_mgprocdir
-    )
+    call_instruction = \
```
Comment (johannbrehmer) [inline madminer/utils/interfaces/mg.py:520]: See https://github.com/diana-hep/madminer/pull/438/files#r512226711

**Review by johannbrehmer — APPROVED** [review]
What a heroic effort! Thank you!

**Review by johannbrehmer — APPROVED** [review]

File: madminer/utils/interfaces/mg.py (hunk 3 above)
Comment (Sinclert) [reply madminer/utils/interfaces/mg.py:494]: Ok. Changing it to `( ... )` style.

File: madminer/delphes/delphes_reader.py (hunk 1 above)
Comment (johannbrehmer) [reply madminer/delphes/delphes_reader.py:865]: Good point, an easy win indeed!

**Comment by johannbrehmer** [comment]
Just for the record, we tried to maintain black code style already before, but with a maximum line length of 120 instead of the default of 88.

I do not have a strong opinion on the line length question, but a slight preference for 120 characters on modern, big screens.

Let me know what you think on this, and then I'll either merge this version or a version with 120 character limit.

File: madminer/utils/interfaces/mg.py (hunk 4 above)
Comment (Sinclert) [reply madminer/utils/interfaces/mg.py]: Ok. Changing it to `( ... )` style.

**Comment by Sinclert** [comment]
I would say increasing the line length over 88 is a practical move, although 120 characters may be too much.

On this PR the max line length has been respected. It may be worth discussing about on a separated issue / PR.

Ready to merge ✅
